File: 113. Path Sum II/Java/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        class Utils {
            void dfs(TreeNode root, int sum, List<List<Integer>> result, List<Integer> current) {
                current.add(root.val);
                sum -= root.val;
                if (sum == 0) {
                    if (root.left == null && root.right == null) {
                        result.add(new ArrayList<>(current));
                    }
                }
                if (root.left != null) {
                    dfs(root.left, sum, result, current);
                }
                if (root.right != null) {
                    dfs(root.right, sum, result, current);
                }
                current.remove(current.size() - 1);
            }
        }
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Utils utils = new Utils();
        utils.dfs(root, sum, result, new ArrayList<>());
        return result;
    }
}

File: 113. Path Sum II/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode *root, int sum) {
        class Utils {
        public:
            void dfs(TreeNode *root, int sum, vector<vector<int>> &result, vector<int> &current) {
                current.push_back(root->val);
                sum -= root->val;
                if (sum == 0) {
                    if (root->left == nullptr && root->right == nullptr) {
                        result.push_back(current);
                    }
                }
                if (root->left != nullptr) {
                    dfs(root->left, sum, result, current);
                }
                if (root->right != nullptr) {
                    dfs(root->right, sum, result, current);
                }
                current.pop_back();
            }
        };
        vector<vector<int>> result;
        if (root == nullptr) {
            return result;
        }
        Utils utils;
        vector<int> temp;
        utils.dfs(root, sum, result, temp);
        return result;
    }
};